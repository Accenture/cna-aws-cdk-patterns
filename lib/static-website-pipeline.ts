import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { Artifact, Pipeline, StageProps } from '@aws-cdk/aws-codepipeline';
import { BitBucketSourceAction, CodeBuildAction, S3DeployAction } from '@aws-cdk/aws-codepipeline-actions';
import { IBucket } from '@aws-cdk/aws-s3';
import { BuildSpec, ComputeType, IProject, LinuxBuildImage, PipelineProject } from '@aws-cdk/aws-codebuild';

export enum StaticWebsitePipelineType {
  SIMPLE,
  ANGULAR
}

export interface StaticWebsitePipelineProps {
  type: StaticWebsitePipelineType,
  pipelineName: string,
  bucket: IBucket,
  codeStarConnectionArn: string,
  repository: {
    owner: string,
    name: string,
    branch?: string
  }
}

export class StaticWebsitePipeline extends Construct {
  constructor(scope: Construct, id: string, props: StaticWebsitePipelineProps) {
    super(scope, id);

    let stages = this.simplePipelineStages(props);
    if(props.type === StaticWebsitePipelineType.ANGULAR) {
      stages = this.angularPipelineStage(props);
    }

    new Pipeline(this, 'pipeline', {
        pipelineName: props.pipelineName,
        crossAccountKeys: false,
        stages: stages
    });
  }

  simplePipelineStages(props: StaticWebsitePipelineProps): StageProps[] {
    const sourceArtifact = new Artifact('Source');

    return [
      {
        stageName: 'Source',
        actions: [
          new BitBucketSourceAction({
            actionName: 'Checkout',
            connectionArn: props.codeStarConnectionArn,
            owner: props.repository.owner,
            repo: props.repository.name,
            branch: props.repository.branch,
            output: sourceArtifact
          })
        ]
      },
      {
        stageName: 'Deploy',
        actions: [
          new S3DeployAction({
            actionName: 'Deploy',
            bucket: props.bucket,
            input: sourceArtifact
          })
        ]
      }
    ]
  }

  angularPipelineStage(props: StaticWebsitePipelineProps): StageProps[] {
    const sourceArtifact = new Artifact('Source');
    const deployArtifact = new Artifact('Deploy');

    return [
      {
        stageName: 'Source',
        actions: [
          new BitBucketSourceAction({
            actionName: 'Checkout',
            connectionArn: props.codeStarConnectionArn,
            owner: props.repository.owner,
            repo: props.repository.name,
            branch: props.repository.branch,
            output: sourceArtifact
          })
        ]
      },
      {
        stageName: 'Build',
        actions: [
          new CodeBuildAction({
            actionName: 'Build',
            project: this.angularBuildProject(),
            input: sourceArtifact,
            outputs: [ deployArtifact ]
          })
        ]
      },
      {
        stageName: 'Deploy',
        actions: [
          new S3DeployAction({
            actionName: 'Deploy',
            bucket: props.bucket,
            input: deployArtifact
          })
        ]
      }
    ]
  }

  angularBuildProject(): IProject {
    return new PipelineProject(this, 'build-project', {
      projectName: 'AngularAppBuild',
      environment: {
        computeType: ComputeType.SMALL,
        buildImage: LinuxBuildImage.STANDARD_4_0
      },
      buildSpec: BuildSpec.fromObject({
        version: '0.2',
        phases: {
          install: {
            'runtime-versions': {
              nodejs: 12
            },
            commands: [
              'npm install -g @angular/cli@10',
              'npm install'
            ]
          },
          build: {
            commands: [
              'ng build --prod=true --outputPath=dist'
            ]
          }
        },
        artifacts: {
          'base-directory': 'dist',
          files: [
            '**/*'
          ]
        }
      })
    });
  }
}
