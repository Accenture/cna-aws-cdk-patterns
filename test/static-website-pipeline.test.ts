import * as cdk from '@aws-cdk/core';
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { StaticWebsitePipeline, StaticWebsitePipelineType } from '../lib/static-website-pipeline';
import { Bucket } from '@aws-cdk/aws-s3';

test('Simple pipeline has two stages', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    // WHEN
    new StaticWebsitePipeline(stack, 'test-website-pipeline', {
        type: StaticWebsitePipelineType.SIMPLE,
        pipelineName: 'test-pipeline',
        bucket: new Bucket(stack, 'bucket'),
        codeStarConnectionArn: '',
        repository: {
            owner: 'test',
            name: 'website'
        }
    });
    // THEN
    expect(stack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Name: 'test-pipeline',
        Stages: [
            {
                Name: 'Source',
                Actions: [
                    {
                        Name: 'Checkout'
                    }
                ]
            },
            {
                Name: 'Deploy',
                Actions: [
                    {
                        Name: 'Deploy'
                    }
                ]
            }
        ]
    });

    // In addition to the fine granular assertions snapshot tests are performed, too
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Angular pipeline has three stages', () => {
    const app = new cdk.App();
    const stack = new cdk.Stack(app, 'TestStack');
    // WHEN
    new StaticWebsitePipeline(stack, 'test-website-pipeline', {
        type: StaticWebsitePipelineType.ANGULAR,
        pipelineName: 'test-pipeline',
        bucket: new Bucket(stack, 'bucket'),
        codeStarConnectionArn: '',
        repository: {
            owner: 'test',
            name: 'website'
        }
    });
    // THEN
    expect(stack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Name: 'test-pipeline',
        Stages: [
            {
                Name: 'Source',
                Actions: [
                    {
                        Name: 'Checkout'
                    }
                ]
            },
            {
                Name: 'Build',
                Actions: [
                    {
                        Name: 'Build'
                    }
                ]
            },
            {
                Name: 'Deploy',
                Actions: [
                    {
                        Name: 'Deploy'
                    }
                ]
            }
        ]
    });

    // In addition to the fine granular assertions snapshot tests are performed, too
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
