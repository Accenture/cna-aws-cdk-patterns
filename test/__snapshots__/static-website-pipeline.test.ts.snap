// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Angular pipeline has three stages 1`] = `
Object {
  "Resources": Object {
    "bucket43879C71": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testwebsitepipeline2DE54661": Object {
      "DependsOn": Array [
        "testwebsitepipelineRoleDefaultPolicyD2B93626",
        "testwebsitepipelineRole1737B58B",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "testwebsitepipelineArtifactsBucket3773A994",
          },
          "Type": "S3",
        },
        "Name": "test-pipeline",
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "testwebsitepipelineRole1737B58B",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": Object {
                  "BranchName": "master",
                  "ConnectionArn": "",
                  "FullRepositoryId": "test/website",
                },
                "Name": "Checkout",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "Source",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Build",
                  "Owner": "AWS",
                  "Provider": "CodeBuild",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ProjectName": Object {
                    "Ref": "testwebsitepipelinebuildproject37A558CC",
                  },
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "Source",
                  },
                ],
                "Name": "Build",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "Deploy",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineBuildCodePipelineActionRole42EA6D1B",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Build",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1",
                },
                "Configuration": Object {
                  "BucketName": Object {
                    "Ref": "bucket43879C71",
                  },
                  "Extract": "true",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "Deploy",
                  },
                ],
                "Name": "Deploy",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Deploy",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "testwebsitepipelineArtifactsBucket3773A994": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testwebsitepipelineBuildCodePipelineActionRole42EA6D1B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineBuildCodePipelineActionRoleDefaultPolicyF1A6FFEC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "codebuild:BatchGetBuilds",
                "codebuild:StartBuild",
                "codebuild:StopBuild",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelinebuildproject37A558CC",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineBuildCodePipelineActionRoleDefaultPolicyF1A6FFEC",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineBuildCodePipelineActionRole42EA6D1B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelineDeployCodePipelineActionRole974D5E5C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineDeployCodePipelineActionRoleDefaultPolicyE537695D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "bucket43879C71",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "bucket43879C71",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineDeployCodePipelineActionRoleDefaultPolicyE537695D",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelineRole1737B58B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineRoleDefaultPolicyD2B93626": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelineBuildCodePipelineActionRole42EA6D1B",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineRoleDefaultPolicyD2B93626",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineRole1737B58B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineSourceCheckoutCodePipelineActionRoleDefaultPolicy6E27C879": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "codestar-connections:UseConnection",
              "Effect": "Allow",
              "Resource": "",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineSourceCheckoutCodePipelineActionRoleDefaultPolicy6E27C879",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelinebuildproject37A558CC": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "CODEPIPELINE",
        },
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:4.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "AngularAppBuild",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "testwebsitepipelinebuildprojectRole8469E63A",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "{
  \\"version\\": \\"0.2\\",
  \\"phases\\": {
    \\"install\\": {
      \\"runtime-versions\\": {
        \\"nodejs\\": 12
      },
      \\"commands\\": [
        \\"npm install -g @angular/cli@10\\",
        \\"npm install\\"
      ]
    },
    \\"build\\": {
      \\"commands\\": [
        \\"ng build --prod=true --outputPath=dist\\"
      ]
    }
  },
  \\"artifacts\\": {
    \\"base-directory\\": \\"dist\\",
    \\"files\\": [
      \\"**/*\\"
    ]
  }
}",
          "Type": "CODEPIPELINE",
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "testwebsitepipelinebuildprojectRole8469E63A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelinebuildprojectRoleDefaultPolicy8C630853": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "testwebsitepipelinebuildproject37A558CC",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":log-group:/aws/codebuild/",
                      Object {
                        "Ref": "testwebsitepipelinebuildproject37A558CC",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":report-group/",
                    Object {
                      "Ref": "testwebsitepipelinebuildproject37A558CC",
                    },
                    "-*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelinebuildprojectRoleDefaultPolicy8C630853",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelinebuildprojectRole8469E63A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`Simple pipeline has two stages 1`] = `
Object {
  "Resources": Object {
    "bucket43879C71": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testwebsitepipeline2DE54661": Object {
      "DependsOn": Array [
        "testwebsitepipelineRoleDefaultPolicyD2B93626",
        "testwebsitepipelineRole1737B58B",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "Location": Object {
            "Ref": "testwebsitepipelineArtifactsBucket3773A994",
          },
          "Type": "S3",
        },
        "Name": "test-pipeline",
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "testwebsitepipelineRole1737B58B",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeStarSourceConnection",
                  "Version": "1",
                },
                "Configuration": Object {
                  "BranchName": "master",
                  "ConnectionArn": "",
                  "FullRepositoryId": "test/website",
                },
                "Name": "Checkout",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "Source",
                  },
                ],
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "S3",
                  "Version": "1",
                },
                "Configuration": Object {
                  "BucketName": Object {
                    "Ref": "bucket43879C71",
                  },
                  "Extract": "true",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "Source",
                  },
                ],
                "Name": "Deploy",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Deploy",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "testwebsitepipelineArtifactsBucket3773A994": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testwebsitepipelineDeployCodePipelineActionRole974D5E5C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineDeployCodePipelineActionRoleDefaultPolicyE537695D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "bucket43879C71",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "bucket43879C71",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineDeployCodePipelineActionRoleDefaultPolicyE537695D",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelineRole1737B58B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineRoleDefaultPolicyD2B93626": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testwebsitepipelineDeployCodePipelineActionRole974D5E5C",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineRoleDefaultPolicyD2B93626",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineRole1737B58B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testwebsitepipelineSourceCheckoutCodePipelineActionRoleDefaultPolicy6E27C879": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "codestar-connections:UseConnection",
              "Effect": "Allow",
              "Resource": "",
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "testwebsitepipelineArtifactsBucket3773A994",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "testwebsitepipelineArtifactsBucket3773A994",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testwebsitepipelineSourceCheckoutCodePipelineActionRoleDefaultPolicy6E27C879",
        "Roles": Array [
          Object {
            "Ref": "testwebsitepipelineSourceCheckoutCodePipelineActionRole40A43252",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
